fvm/transaction.go

```diff
@@ -70,65 +76,123 @@ func (i *TransactionInvocator) Process(
 	proc *TransactionProcedure,
 	st *state.State,
 ) error {
-	env, err := newEnvironment(ctx, vm, st)
-	if err != nil {
-		return err
-	}
-	env.setTransaction(proc.Transaction, proc.TxIndex)
 
-	location := common.TransactionLocation(proc.ID[:])
+	var err error
+	var env *hostEnv
+
+	numberOfRetries := 0
+	for numberOfRetries = 0; numberOfRetries < int(ctx.MaxNumOfTxRetries); numberOfRetries++ {
+		env, err = newEnvironment(ctx, vm, st)
+		// env construction error is fatal
+		if err != nil {
+			return err
+		}
+		env.setTransaction(proc.Transaction, proc.TxIndex)
+
+		location := common.TransactionLocation(proc.ID[:])
+
+		err = vm.Runtime.ExecuteTransaction(
+			runtime.Script{
+				Source:    proc.Transaction.Script,
+				Arguments: proc.Transaction.Arguments,
+			},
+			runtime.Context{
+				Interface: env,
+				Location:  location,
+			},
+		)
+
+		// break the loop
+		if !i.requiresRetry(err, proc) {
+			break
+		}
+
+		var blockHeight uint64
+		if ctx.BlockHeader != nil {
+			blockHeight = ctx.BlockHeader.Height
+		}
 
-	err = vm.Runtime.ExecuteTransaction(
-		runtime.Script{
-			Source:    proc.Transaction.Script,
-			Arguments: proc.Transaction.Arguments,
-		},
-		runtime.Context{
-			Interface: env,
-			Location:  location,
-		},
-	)
+		i.logger.Info().
+			Str("txHash", proc.ID.String()).
+			Uint64("blockHeight", blockHeight).
+			Int("retries_count", numberOfRetries).
+			Uint64("ledger_interaction_used", st.InteractionUsed()).
+			Msg("retrying transaction execution")
+
+		// reset error part of proc
+		// Warning right now the tx requires retry logic doesn't change
+		// anything on state but we might want to revert the state changes (or not commiting)
```

Comment:
so we don't reset the state currently? how much extra work is it to throw out the state changes? Just in case we do run into a case where the state does change. (we want to fix the bug where failed tx doesn't change state)

Reply:
Right now we don't reset the state writes since we only retry on checking failed transactions, I agree we should change the logic to prevent future mistakes in the code but that's a fair amount of work and we need to solve the bug you mentioned first. I'll address it in the PR addressing the bug.

Reply:
the current issue is if I rollback here seq number updates will be wiped for this tx :(